# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Time & Space Complexity</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a)How do we measure it?<br>
8) <b>In/Out-Place Algorithms</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Out of Place Algorithm <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) In-Class Exercise #1 <br>
9) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>


## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def addTwo(x):
    return x + 2

print(addTwo(4))

#Lambda Function Syntax
# Example of calling a Lambda function withot a variable
print((lambda x:x+2)(4))


6
6


#### Saving to a Variable

In [5]:
f_test = lambda x: x + 2  

f_test(4)    

6

#### Multiple Inputs

In [10]:
#multiple Inputs with no variable
print((lambda x,y,z: x * y * z)(3,5,8))

#Muliple Inputs with a variable attached
x_test = lambda x,y,z: x*y*z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [13]:
def multiply(f,num):
    """
    f expects a lambda function num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [17]:
#Regular defined function
def multiply_test(num):
    return num * 4
#Function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply


f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda functon returned from a regular function
def returnLam(b,c)



<function returnFunc.<locals>.multiply at 0x000001BE6A48CF70>
8


#### If Statements within Lambdas

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [ ]:
nums = {5, 7, 9, 12, 30}





#### Using Lambda's with Map

In [1]:
nums = [3, 0, 11, 33, 50]
result = list(map(lambda num: num**2, nums))
print(result)

#map(func, iter)

[9, 0, 121, 1089, 2500]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [3]:
nums = [3, 0, 11, 33, 50]
result = list(map(lambda num: num*2 -1, nums))
print(result)

[5, -1, 21, 65, 99]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [ ]:
#filter(func, iter)
names = [Bobe]
def b-name(name):
    if name[0].lower == 'b':
        return True
    else:
        return False

result = list(filter(b_name, name))
print(result)

#### Using Lambda's with Filter()

In [14]:
result = list(filter(lambda name:True if name[0].lower() == 'b', name )
print(result)

SyntaxError: expected 'else' after 'if' expression (1038213541.py, line 1)

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [17]:
nums = [2,7,4.2,1.6,9,4.4,4.9]
result = list(filter(lambda num: True if num > sum(nums)/len(nums)else False, nums))
print(result)


[7, 9, 4.9]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [19]:
#reduce(func, iter)
from functools import reduce

num = [2,4,6,8]
def add_nums(p1,p2):
    return p1 + p2

result = reduce(add_nums, nums)
print(result)

33.099999999999994


#### Using Lambda's with Reduce()

In [20]:
reduce(lambda x,y: x+y [3,423,45,4334])
print(result)

TypeError: reduce expected at least 2 arguments, got 1

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [50]:
def add_nums(num):
    #set a base case
    if num == 0:
        return 0
    #recursive case
    else:
        return num + add_nums(num-1)
add_num(5)

NameError: name 'add_num' is not defined

#### Writing a Factorial Function

In [49]:
def factorial(num):
    if num <=1:
        return 1
    else:
        return num * factorial(num -1)
factorial(5)    

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [ ]:
def my_range(start, stop = None, step = 1)
    if stop is None:
        stop = start
        start = 0
    output = []
    while start < stop:
        yield start
        start += step
        
nums = my_range(3,10,2)
print(nums)
for num in nums:
    print(num)

#### Infinite Generator

In [ ]:
# bad, never create infinite loops


#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
# for i in numbers_squared(10)

def num_sq(stop,start, step = 1):
    while start >= stop:
        yield start **2
        

## Time and Space Complexity

#### What is it?

Time and space complexity is the measure of how much time a given action(function) will take to solve a problem. In the same fashion, we determine how much a given data structure will need in terms of memory allocation. A problem can have multiple solutions and finding the optimal solution for the problem needs to be analyzed in time and space.

#### How do we measure Time and Space Complexity?

In order to measure time and space complexity we use Asymptotic analysis. The reason for this is because we need a way to measure different algorithms (functions) based on the size of their inputs in a mathmatical way. For example, we could have a function that is computed as f(n) and another that is g(n^2). All things around the function staying constant, the only thing that changes is the size of the input. Below is the chart that shows the different Asymptotic analysis formats. 

<table style="text-align:center;" class="table table-bordered">
<tbody><tr>
<td>constant</td>
<td>−</td>
<td>Ο(1)</td>
</tr>
<tr>
<td>logarithmic</td>
<td>−</td>
<td>Ο(log n)</td>
</tr>
<tr>
<td>linear</td>
<td>−</td>
<td>Ο(n)</td>
</tr>
<tr>
<td>Linear Logarithmic</td>
<td>−</td>
<td>Ο(n log n)</td>
</tr>
<tr>
<td>quadratic</td>
<td>−</td>
<td>Ο(n<sup>2</sup>)</td>
</tr>
<tr>
<td>cubic</td>
<td>−</td>
<td>Ο(n<sup>3</sup>)</td>
</tr>
<tr>
<td>polynomial</td>
<td>−</td>
<td>n<sup>Ο(1)</sup></td>
</tr>
<tr>
<td>exponential</td>
<td>−</td>
<td>2<sup>Ο(n)</sup></td>
</tr>
</tbody></table>

## In Place Algorithms

#### Syntax

In [ ]:
def swap(nums, x,y):
    nums[x], nums[y]=nums[y],nums[x]
    
my-list = {4,11,12,22,100,64}
print(f'Before swap: {my-list}')
    

#### Out of Place Algorithm

#### In-Class Exercise <br>
<p>Write a function that takes in four arguments (list, index1, index2, index3), and swaps those three positions in the list passed in.</p>

In [ ]:
def swap(num)

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [63]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
#while " " in places:
    #places.remove(" ")
#print(places)    
#while "" in places:
    #places.remove("")
#print(places)
#while "  " in places:
   # places.remove("  ")
#print(places)    


result = list(filter(lambda place: place.strip() !="",places))
print(result)
    #now white spaces

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [74]:
#list.sort
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
#print(sorted(author))
#acces key
#print(author[0])
#print(author[1])
#print(author[2])
#print(author[3])
#print(author[4])
author.sort(key = (lambda name: name.lower().split()[-1]))
print(author)
#for a in author:
    #print(a.split()[-1])


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [76]:
## nums = [3, 0, 11, 33, 50]
#result = list(map(lambda num: num**2, nums))
#print(result)
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
result = list(map(lambda place: (place[0], float(place[1])*(9/5)+32), places))
print(result)                  

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [4]:
# 5! = 5 * 4 * 3 * 2 * 1
#def factorial(num):
   # if num <= 1:
     #   return 1
  #  else:
      #  return num * factorial(num - 1)
def fib(num):
    if num <= 1:
        return 1
    else:
        return fib(num-1) + fib(num-2)

def pfib(num):
    for i in range(num +1):
        print(f'inderation {i}: {fib(i)}')
pfib(5)


inderation 0: 1
inderation 1: 1
inderation 2: 2
inderation 3: 3
inderation 4: 5
inderation 5: 8
